compiler/rustc_codegen_ssa/src/base.rs

```diff
@@ -154,7 +154,9 @@ pub fn unsized_info<'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>>(
         ) if src_dyn_kind == target_dyn_kind => {
             let old_info =
                 old_info.expect("unsized_info: missing old info for trait upcasting coercion");
-            if data_a.principal_def_id() == data_b.principal_def_id() {
+            if data_a.principal_def_id() == data_b.principal_def_id()
+                || data_b.principal().is_none()
```

Comment:
@bjorn3: if this PR ends up being accepted, the change will need to be made to cg-clif as well. should I do that in this PR, or in a separate one?

Reply:
Doing it in the same PR would be nice I think. You can change https://github.com/rust-lang/rust/blob/a32978a5e8a3c34bb5caa1175f7404c2ede019b6/compiler/rustc_codegen_cranelift/src/unsize.rs#L34-L37 the same way as you did here.